src/ray/common/asio/asio_chaos.cc

```diff
@@ -0,0 +1,136 @@
+
+   Please check warnings to make sure delay is on.
+*/
+class DelayManager {
+ public:
+  DelayManager() { Init(); }
+
+  int64_t GetMethodDelay(const std::string &name) const {
+    auto it = delay_.find(name);
+    if (it == delay_.end()) {
+      return GenRandomDelay(random_delay_us_.first, random_delay_us_.second);
+    }
+    int64_t actual_delay = GenRandomDelay(it->second.first, it->second.second);
+    if (actual_delay != 0) {
+      RAY_LOG_EVERY_N(ERROR, 1000)
+          << "Delaying method " << name << " for " << actual_delay << "us";
+    }
+    return actual_delay;
+  }
+
+  void Init() {
+    delay_.clear();
+    random_delay_us_ = {0, 0};
+    auto delay_env = RayConfig::instance().testing_asio_delay_us();
+    if (delay_env.empty()) {
+      return;
+    }
+    RAY_LOG(ERROR) << "RAY_testing_asio_delay_us is set to " << delay_env;
+    std::vector<std::string_view> items = absl::StrSplit(delay_env, ",");
+    for (const auto &item : items) {
+      ParseItem(item);
+    }
+  }
+
+ private:
+  // method1=min:max
+  void ParseItem(std::string_view val) {
+    std::vector<std::string_view> item_val = absl::StrSplit(val, "=");
+    if (item_val.size() != 2) {
+      RAY_LOG(ERROR) << "Error in syntax: " << val
+                     << ", expected method=min_us:max:ms. Skip this entry.";
+    }
+    auto delay_us = ParseVal(item_val[1]);
+    if (delay_us) {
+      if (item_val[0] == "*") {
+        random_delay_us_ = *delay_us;
+      } else {
+        delay_[item_val[0]] = *delay_us;
+      }
+    }
+  }
+
+  std::optional<std::pair<int64_t, int64_t>> ParseVal(std::string_view val) {
+    std::vector<std::string_view> delay_str_us = absl::StrSplit(val, ":");
+    if (delay_str_us.size() != 2) {
+      RAY_LOG(ERROR) << "Error in syntax: " << val
+                     << ", expected method=min_us:max:ms. Skip this entry";
+      return std::nullopt;
+    }
+    std::pair<int64_t, int64_t> delay_us;
+    if (!absl::SimpleAtoi(delay_str_us[0], &delay_us.first) ||
+        !absl::SimpleAtoi(delay_str_us[1], &delay_us.second)) {
+      RAY_LOG(ERROR) << "Error in syntax: " << val
+                     << ", expected method=min_us:max:ms. Skip this entry";
+      return std::nullopt;
+    }
+    if (delay_us.first > delay_us.second) {
+      RAY_LOG(ERROR) << delay_us.first << " is bigger than " << delay_us.second
+                     << ". Skip this entry.";
+      return std::nullopt;
+    }
+    return delay_us;
+  }
+
+  int64_t GenRandomDelay(int64_t min_delay_us, int64_t max_delay_us) const {
+    if (min_delay_us == max_delay_us) {
+      return min_delay_us;
+    }
+    return std::rand() % (max_delay_us - min_delay_us) + min_delay_us;
+  }
+
+  absl::flat_hash_map<std::string, std::pair<int64_t, int64_t>> delay_;
+  std::pair<int64_t, int64_t> random_delay_us_ = {0, 0};
+};
+
+static DelayManager _delay_manager;
+}  // namespace
+
+int64_t get_delay_us(const std::string &name) {
+  return _delay_manager.GetMethodDelay(name);
```

Comment:
Won't this cause a hash table lookup even if delays are disabled?

Reply:
it'll, but given the hash table is empty, I think it's ok?

Reply:
You're still adding an extra string hash operation to every asio operation. Could we flag this with a boolean?